### PR TITLE
IP_COOLDOWN_PERIOD environment variable for ip cooldown period configuration

### PR DESCRIPTION
**Pull request by jchen6585**
**What type of PR is this?**
Continuing PR #2397 

**Which issue does this PR fix**:
Issue #2378 

**What does this PR do / Why do we need it**:
Continuing PR #2397: Added new environment variable IP_COOLDOWN_PERIOD.  cmd/aws-vpc-cni/main.go defines two new variables in the const section, envIPCooldownPeriod with IP_COOLDOWN_PERIOD and defaultIPCooldownPeriod as 30s time.Duration. pkg/ipamd/datastore/data_store.go creates a new function getCooldownPeriod(), returns a time.Duration being either the default cooldown period of 30s or the custom one if it is not smaller than 0. getCooldownPeriod() is called when creating a new DataStore, storing the cooldown period value in a variable ipCooldownPeriod which is used in inCoolingPeriod(). pkg/ipamd/datastore/data_store_test.go creates a new test case TestGetIPStatsV4WithCustomIPCooldown(). util/utils.go adds a new function GetIntAsStringEnvVar() to parse the environment string into an integer.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

Ran against cni and ipamd integration tests with IP_COOLDOWN_PERIOD values set as: unset, 0, 10, 20
Screenshot 2023-08-09 at 2 25 25 PM
Screenshot 2023-08-09 at 2 26 11 PM

**Automation added to e2e**:

No

**Will this PR introduce any new dependencies?**:

No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
Have not updated a running cluster

**Does this change require updates to the CNI daemonset config files to work?**:

No

**Does this PR introduce any user-facing change?**:

Yes

```release-note
#### `IP_COOLDOWN_PERIOD` (v1.13.4+)

Type: Integer as a String

Default: `30`

Specifies the number of seconds an IP address is in cooldown after pod deletion. The cooldown period gives kube-proxy time to update node iptables rules when the IP was registered as a valid endpoint, such as for a service. Modify this value with caution, as kube-proxy update time scales with the number of nodes and services.

**Note:** 0 is a supported value, however it is highly discouraged.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.md
Comment (jdn5126) [inline README.md:685]: This will end up being `v1.14.1+`, as we will have to hold off on merging this until the `v1.14.0` release is complete

File: README.md
Comment (jdn5126) [inline README.md:691]: nit: let's change "gives kube-proxy" to "gives network proxies, such as kube-proxy,"

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:357]: nit: let's move this above the variables defined above. Otherwise it makes that logic harder to read

File: utils/utils.go
Comment (jdn5126) [inline utils/utils.go:25]: nit: maybe `GetIntFromStringEnvVar`?

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:360]: For both error cases, let's print the value that the user passed

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:284]: should be no need for `time.Duration` here. It is just an integer times `time.Second`

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:369]: I still think we should just make this as a member variable of `DataStore`

**Review by jdn5126 — CHANGES_REQUESTED** [review]

File: pkg/ipamd/datastore/data_store.go
Comment (jchen6585) [reply pkg/ipamd/datastore/data_store.go:280]: Removing time.Duration results in this error `invalid operation: cooldownVal * time.Second (mismatched types int and time.Duration)`

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:40]: nit: let's clean up this comment to say: "envIPCooldownPeriod (default 30 seconds) specifies the time after pod deletion before an IP can be assigned to a new pod"

File: pkg/ipamd/datastore/data_store_test.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store_test.go:928]: nit: this comment should remain, as it still applies

File: pkg/ipamd/datastore/data_store_test.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store_test.go:979]: this comment should remain, as it still applies

File: pkg/ipamd/datastore/data_store_test.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store_test.go:928]: We should take this opportunity to fix this test, as sleeping for 30 seconds is not ideal. Maybe we can set the cooldown period lower to start? Or change the time

**Review by jdn5126 — COMMENTED** [review]
A few comments, this is almost there

File: utils/utils.go
Comment (jdn5126) [inline utils/utils.go:32]: nit: let's return -1 in the failure case

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:355]: nit: Let's combine this case and the case below, i.e. `if err != nil || ipCooldownPeriod < 0`.

Since this is a container log, let's change this to "IP_COOLDOWN_PERIOD MUST be a valid positive integer. %s is invalid"

**Review by jdn5126 — CHANGES_REQUESTED** [review]
Sorry for the delay, these are the last nits, then this is good to go! Let's also update the title and clean up the description a bit for future readers.

File: utils/utils.go
Comment (jdn5126) [reply utils/utils.go:32]: Let's also add a small unit test for this in `utils_test.go`

**Comment by jayanthvn** [comment]
Can you add tests here - https://github.com/aws/amazon-vpc-cni-k8s/tree/master/test/integration/ipamd before merging the PR..Also callout increasing the cooldown timer will cause higher EC2 calls since IPs will be in cache for more time.

**Comment by jdn5126** [comment]
> Can you add tests here - https://github.com/aws/amazon-vpc-cni-k8s/tree/master/test/integration/ipamd before merging the PR..Also callout increasing the cooldown timer will cause higher EC2 calls since IPs will be in cache for more time.

@jayanthvn I am not sure that an integration test case adds much value here, as the number itself is really arbitrary. We have only been using 30 seconds as a default because that's what upstream Kubernetes recommended early on. If we run the service connectivity test with 10 second cooldown and it passes consistently, that isn't going to change our default, so  this overhead doesn't seem worth it.

@jchen6585 the callout for the README is a good point. More IPs in cooldown means more IPs need to be allocated, which means more EC2 API calls

**Review by jdn5126 — APPROVED** [review]

**Comment by jayanthvn** [comment]
But IMO, I feel we need to ensure DS functionality is fine i.e, no of IPs/total cooldown or at least the IP/ENIs allocated are expected with the setting..also if some change goes in breaking this functionality we should be able to capture it which UTs will not do..

**Comment by jchen6585** [comment]
> But IMO, I feel we need to ensure DS functionality is fine i.e, no of IPs/total cooldown or at least the IP/ENIs allocated are expected with the setting..also if some change goes in breaking this functionality we should be able to capture it which UTs will not do..
> 
> @jchen6585 - Do add our recommendation and reason of 30s in EKS best practices..

@jayanthvn is this what you are referring to? https://aws.github.io/aws-eks-best-practices/networking/index/
if so, how do i get access to edit it?

**Comment by jayanthvn** [comment]
> > But IMO, I feel we need to ensure DS functionality is fine i.e, no of IPs/total cooldown or at least the IP/ENIs allocated are expected with the setting..also if some change goes in breaking this functionality we should be able to capture it which UTs will not do..
> > @jchen6585 - Do add our recommendation and reason of 30s in EKS best practices..
> 
> @jayanthvn is this what you are referring to? https://aws.github.io/aws-eks-best-practices/networking/index/ if so, how do i get access to edit it?

You can open a request here - https://github.com/aws/aws-eks-best-practices

**Review by jdn5126 — APPROVED** [review]

**Pull request merged**
